File: task_1/go/1(2).go
package main

import (
	"fmt"
	"math"
)

// φ1(x) для положительного корня
func phiPositive(x float64) float64 {
	return 1.0 / math.Log(x+1)
}

func dphiPositive(x float64) float64 {
	return -1.0 / ((x + 1) * math.Pow(math.Log(x+1), 2)) // производная φ1(x)
}

// φ2(x) для отрицательного корня
func phiNegative(x float64) float64 {
	return math.Exp(1.0/x) - 1
}

func dphiNegative(x float64) float64 {
	return -math.Exp(1.0/x) / (x * x) // производная φ2(x)
}

func iterationMethod(x0, eps float64, phi, dphi func(float64) float64) float64 {
	k := 0
	prevX := x0 - 2*eps

	if math.Abs(dphi(x0)) >= 1 {
		fmt.Println("Метод не сходится при данном x0. Попробуйте другое начальное приближение.")
		return 0.0
	}

	fmt.Println("╔════════════════════════════════════════════════════════════════════════════╗")
	fmt.Println("║   N   ║         x_n           ║       x_(n+1)         ║   x_(n+1) - x_n    ║")
	fmt.Println("╠════════════════════════════════════════════════════════════════════════════╣")

	for math.Abs(x0-prevX) >= eps {
		prevX = x0
		x0 = phi(prevX)

		fmt.Printf("║  %3d  ║\t%.6f\t║\t%.6f\t║\t%.6f     ║\n", k, prevX, x0, math.Abs(x0-prevX))

		k++
	}

	fmt.Println("╚════════════════════════════════════════════════════════════════════════════╝")
	return x0
}

func main() {
	eps := 1e-4

	fmt.Println("\n=== Поиск положительного корня ===")
	x0Pos := 1.5
	rootPos := iterationMethod(x0Pos, eps, phiPositive, dphiPositive)
	fmt.Printf("\nПоложительный корень: x ≈ %.6f\n", rootPos)

	fmt.Println("\n=== Поиск отрицательного корня ===")
	x0Neg := -0.5
	rootNeg := iterationMethod(x0Neg, eps, phiNegative, dphiNegative)
	fmt.Printf("\nОтрицательный корень: x ≈ %.6f\n", rootNeg)
}

File: task_1/cpp/1(2).cpp
#include <iostream>
#include <cmath>
#include <iomanip>
using namespace std;

// φ1(x) для положительного корня
double phi_positive(double x) {
    return 1.0 / log(x + 1);
}

double dphi_positive(double x) {
    return -1.0 / ((x + 1) * pow(log(x + 1), 2)); // производная φ1(x)
}

// φ2(x) для отрицательного корня
double phi_negative(double x) {
    return exp(1.0 / x) - 1;
}

double dphi_negative(double x) {
    return -exp(1.0 / x) / (x * x); // производная φ2(x)
}

double iterationMethod(double x0, double eps,
                       double (*phi)(double), double (*dphi)(double)) {
    int k = 0;
    double prevX = x0 - 2 * eps;

    if (fabs(dphi(x0)) >= 1) {
        cout << "Метод не сходится при данном x0. Попробуйте другое начальное приближение." << endl;
        return 0.0;
    }

    cout << "╔════════════════════════════════════════════════════════════════════════════╗\n";
    cout << "║   N   ║         x_n           ║       x_(n+1)         ║   x_(n+1) - x_n    ║\n";
    cout << "╠════════════════════════════════════════════════════════════════════════════╣\n";

    while (fabs(x0 - prevX) >= eps) {
        prevX = x0;
        x0 = phi(prevX);

        cout << fixed << setprecision(6)
        << "║  " << setw(3) << k << "  ║\t"
        << prevX << "\t║\t" << x0 << "\t║\t"
        << fabs(x0 - prevX) << "     ║" << "\n";

        ++k;
    }

    cout << "╚════════════════════════════════════════════════════════════════════════════╝\n";
    return x0;
                       }

                       int main() {
                           double eps = 1e-4;

                           cout << "\n=== Поиск положительного корня ===\n";
                           double x0_pos = 1.5;
                           double root_pos = iterationMethod(x0_pos, eps, phi_positive, dphi_positive);
                           cout << "\nПоложительный корень: x ≈ " << root_pos << "\n";

                           cout << "\n=== Поиск отрицательного корня ===\n";
                           double x0_neg = -0.5;
                           double root_neg = iterationMethod(x0_neg, eps, phi_negative, dphi_negative);
                           cout << "\nОтрицательный корень: x ≈ " << root_neg << "\n";

                           return 0;
                       }
